iceoryx2-bb/posix/src/periodic_timer.rs

```diff
@@ -0,0 +1,235 @@
+        let start_time = fail!(from "Attachment::new()", when Time::now_with_clock(clock_type),
+                                "Failed to create PeriodicTimer attachment since the current time could not be acquired.");
+        let start_time = start_time.as_duration().as_nanos();
+
+        Ok(Self {
+            index,
+            period,
+            start_time,
+        })
+    }
+
+    fn reset(&mut self, clock_type: ClockType) -> Result<(), TimeError> {
+        let start_time = fail!(from "Attachment::new()", when Time::now_with_clock(clock_type),
+                                "Failed to reset PeriodicTimer attachment since the current time could not be acquired.");
+        self.start_time = start_time.as_duration().as_nanos();
+        Ok(())
+    }
+}
+
+/// The [`PeriodicTimer`] allows the user to attach multiple periodic timers with
+/// [`PeriodicTimer::add()`], to wait on them by acquiring the waiting time to the next timer
+/// with [`PeriodicTimer::next_iteration()`] and to acquire all missed timers via
+/// [`PeriodicTimer::missed_timers()`].
+#[derive(Debug)]
+pub struct PeriodicTimer {
+    timeouts: RefCell<Vec<Attachment>>,
+    id_count: IoxAtomicU64,
+    last_iteration: RefCell<u128>,
+
+    clock_type: ClockType,
+}
+
+impl PeriodicTimer {
+    /// Adds a cyclic timeout to the [`PeriodicTimer`] and returns an [`PeriodicTimerGuard`] to
+    /// identify the attachment uniquely.
+    /// [`PeriodicTimer::next_iteration()`] will schedule the timings in a way that the attached
+    /// timeout is considered cyclicly.
+    pub fn cyclic(&self, timeout: Duration) -> Result<PeriodicTimerGuard, TimeError> {
+        let current_idx = self.id_count.load(Ordering::Relaxed);
+        self.timeouts.borrow_mut().push(Attachment::new(
+            current_idx,
+            timeout.as_nanos(),
+            self.clock_type,
+        )?);
+        self.id_count.fetch_add(1, Ordering::Relaxed);
+
+        Ok(PeriodicTimerGuard {
+            periodic_timer: self,
+            index: current_idx,
+        })
+    }
+
+    fn remove(&self, index: u64) {
+        let mut index_to_remove = None;
+        for (n, attachment) in self.timeouts.borrow().iter().enumerate() {
+            if attachment.index == index {
+                index_to_remove = Some(n);
+                break;
+            }
+        }
+
+        if let Some(n) = index_to_remove {
+            self.timeouts.borrow_mut().remove(n);
+        }
+    }
+
+    fn reset(&self, index: u64) -> Result<(), TimeError> {
+        for attachment in &mut *self.timeouts.borrow_mut() {
+            if attachment.index == index {
+                attachment.reset(self.clock_type)?;
+                break;
+            }
+        }
+
+        Ok(())
+    }
+
+    /// Returns the waiting duration until the next added timeout is reached.
+    pub fn next_iteration(&self) -> Result<Duration, TimeError> {
```

Comment:
```suggestion
    pub fn time_to_next(&self) -> Result<Duration, TimeError> {
```

Thoughts?

Also, what is the goal with this function?

Reply:
The new name `duration_until_next_timeout` is good.